### PR TITLE
[REVIEW FOR REVIEW] nymjs

### PR DESCRIPTION
**Pull request by lsankar4033**
cc @amirgamil @cha0sg0d 

see [here](https://github.com/personaelabs/nym/blob/lsankar/prover-js/packages/circuits/tests/test_utils.ts#L55) for WIP circuit input prep code

output of this PR will be proving frontend that takes `ethers` + `ethereumjs`-like objects (i.e. what we get from wagmi) and generates proofs

### CODE REVIEW

File: packages/circuits/tests/nym_ownership.test.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+describe("nym ownership", () => {
+  it("e2e test with signatures from web frontend", async () => {
+    const testContent = "nyms are eternal";
+    const testNym = "satoshi";
+
+    const nymInput = await computeNymInput(testSigner, testContent, testNym);
+
+    console.log(nymInput);
+
+    expect(1).toBe(1);
```
Comment (cha0sg0d) [inline packages/circuits/tests/nym_ownership.test.ts:12]: I do the same for initial test construction 🤣

File: packages/nymjs/src/prover.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
+// NOTE: this file should be a reference for frontend code
```
Comment (cha0sg0d) [inline packages/nymjs/src/prover.ts:1]: Noted!

File: packages/nymjs/src/prover.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import { hashPersonalMessage, fromRpcSig } from "@ethereumjs/util";
+
+import { computeEffEcdsaPubInput } from "@personaelabs/spartan-ecdsa";
+
+import { buildPoseidon } from "circomlibjs";
+
+import { MerkleProof, NymInput, EffECDSASig } from "./types";
+
+function bufferToBigInt(buf: Buffer): bigint {
```
Comment (cha0sg0d) [inline packages/nymjs/src/prover.ts:10]: @lsankar4033 will we eventually upstream these functions into one package? I'm seeing their re-use in `circuits/test/test_utils.ts`

Comment (cha0sg0d) [reply packages/nymjs/src/prover.ts:10]: Cool. I like duplicating functions as needed until we see where things can be re-used + upstreamed.

File: packages/circuits/tests/test_utils.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  content: string,
+  contentSigStr: string,
+
+  nym: string,
+  nymSigStr: string
+): Promise<NymInput> {
+  const nymSig = await computeEffECDSASig(nymSigStr, nym);
+  const contentSig = await computeEffECDSASig(contentSigStr, content);
+
+  // TODO: type-check poseidon input properly once I'm testing proving
```
[the remark this answers is not in the document]
Comment (lsankar4033) [reply packages/circuits/tests/test_utils.ts:80]: I think that'd be ideal

**Comment by lsankar4033** [comment]
@DanTehrani I've been adapting the code you wrote at https://github.com/personaelabs/spartan-ecdsa/tree/main/packages/lib for the `nymjs` package, essentially a lib for frontends to make nym proofs and handle input generation, etc.

I think this approach to libs makes sense b/c it abstracts away proving system details from frontend devs (i.e. if we wnat others to make apps on top of 'nyms' in the future). In this paradigm, we'd be able to change proving system details for a given piece of functionality (i.e. nyms) with releases while keeping the interface for frontends static.

With this in mind, I think it could make sense to split `spartan-ecdsa/packages/lib` into the common utilities for different circuits (i.e. https://github.com/personaelabs/spartan-ecdsa/tree/main/packages/lib/src/wasm and https://github.com/personaelabs/spartan-ecdsa/blob/main/packages/lib/src/helpers/public_input.ts)

and have `nymjs` and `ecdsa-membership` libs that use that under the hood.

Curious what you think about all of this! Lmk if you think this is the wrong approach

**Comment by DanTehrani** [comment]
@lsankar4033  Agree that separating the prover and the circuit utilities will make the stack cleaner.  But I’m not sure if we should do it now; for nymjs I feel like copy-pasting utils code from spartan-ecdsa (and modifying it as required) could be sufficient.

Eventually, we could make the abstractions more clear by having something akin to _gadgets_ in Halo2. But for now, I think we should keep exploring ways of abstraction and see what will make sense for a wider audience.

**Comment by lsankar4033** [comment]
@DanTehrani sounds good! will continue copying over then

**Comment by lsankar4033** [comment]
@DanTehrani taking over merging this in!

TODO:
- TODOs in prover/input preparation
- tests
- publish package

I fleshed out shared types more [here](https://github.com/personaelabs/nym/blob/lsankar/prover-js/packages/nymjs/src/types/self_attesting_content.ts), but nothing more to do here before merging this PR imo

**Comment by DanTehrani** [comment]
- Implement nymjs/.
- Update test_client/ to use nymjs for proving/verification.
  - Proving and verification is done using the hosted circuit and witness generator.
- Add a completeness test for the circuit.


Proving/verification using nymjs:
https://user-images.githubusercontent.com/36762093/234237690-b9cab95b-bdf5-4eac-bea9-30fe791bbb4b.mov
(sorry about the low res; can't upload files over 10MB)

**Comment by DanTehrani** [comment]
@lsankar4033 
I missed implementing the [hash methods](https://github.com/personaelabs/nym/blob/b93ecd095369cf5576a3b04d077ed4a46bc5dc34/packages/nymjs/src/types/self_attesting_content.ts#L21). I will do this soon within this PR!

**Comment by DanTehrani** [comment]
@lsankar4033 
Review for review once again!

1. Implemented id computation 4fbb2db933d754847e95e9cbc8d7a2ef06a86c46
2. [Specced the data model](https://www.notion.so/personae-labs/Nym-data-model-specification-Dan-3f2b53c20b5d4ca89dd7c043463a21a2) and implemented it 6238e6e11ef1d97a9ba628a884123a9305b6d8fa

**Comment by lsankar4033** [comment]
@cha0sg0d check to see if the API for proving makes sense to you too!

**Comment by lsankar4033** [comment]
have a few questions but they're primarily abt data model so i'll add them to https://www.notion.so/personae-labs/Nym-data-model-specification-Dan-3f2b53c20b5d4ca89dd7c043463a21a2 

this PR lgtm though; feel free to squash merge when ready! if we change the data model later, let's treat those changes as new PRs. excellent work here and thanks for picking up my slack @DanTehrani 

@cha0sg0d if you're interested in the API for proving, check out https://github.com/personaelabs/nym/blob/lsankar/prover-js/packages/nymjs/tests/nym.test.ts#L83

**Comment by DanTehrani** [comment]
Thanks for the review @lsankar4033! Will merge

**Pull request merged**